swarm/storage/ldbstore.go

```diff
@@ -227,7 +227,11 @@ func encodeIndex(index *dpaDBIndex) []byte {
 }
 
 func encodeData(chunk *Chunk) []byte {
-	return append(chunk.Key[:], chunk.SData...)
+	// Always create a new underlying array for the returned byte slice.
+	// The chunk.Key array may be used in the returned slice which
+	// may be changed later in the code or by the LevelDB, resulting
+	// that the Key is changed as well.
+	return append(append([]byte{}, chunk.Key[:]...), chunk.SData...)
```

Comment:
@janos I trust you now what you are doing, but why is this necessary? Aren't we copying everything returned by LevelDB into its own memory, before assigning it to `chunk.Key` and `chunk.SData`? How can LevelDB be changing our chunks after they've been retrieved?

Reply:
That may look like, for sure. But, the slice created from the key array may contain the same array, and when it is used in append, the returned slice may reference the same array, or the append can create a new array and copy all bytes into it. If the new array is created, then there in no problem, but if the the same array is used, it reference will be passed to the leveldb, allowing it to modify the slice, actually the shared underlaying array.

Reply:
@janos yes, this is clear. I understand why you are doing this. I guess I am missing where we are using the `encodeFunc`, anyway, it looks fine.

Reply:
Yes, `encodeFunc` is used as `LDBStore.doPut` function as the `LDBStore.encodeDataFunc` to generate the storage data saved under `dkey`. It was needed to be parametrized in LDBStore, for replacing it with the mock store encode data function.